api/secrets.proto

```diff
@@ -0,0 +1,136 @@
+	// - Returns `InvalidArgument` if `RemoveSecretRequest.Name` is empty.
+	// - Returns `NotFound` if the a secret named `RemoveSecretRequest.Name` is not
+	//   found or if it is found but has no version `RemoveSecretRequest.Version`.
+	// - Returns an error if the deletion fails.
+	rpc RemoveSecret(RemoveSecretRequest) returns (RemoveSecretResponse) {
+		option (docker.protobuf.plugin.tls_authorization) = { roles: "swarm-manager" };
+	}
+}
+
+// GetSecretRequest is the request to get a `Secret` object given a secret name.
+message GetSecretRequest {
+	string name = 1;
+}
+
+// GetSecretResponse contains the Secret corresponding to the name in
+// `GetSecretRequest`, but the `SecretData.Spec.Data` field in each `SecretData`
+// object in the `Secret` should be nil instead of actually containing
+// the secret bytes.
+message GetSecretResponse {
+	Secret secret = 1;
+}
+
+// ListSecretRequest is the request to list all secrets in the secret store,
+// or all secrets filtered by name or name prefix.
+message ListSecretsRequest {
+	message Filters {
+		repeated string names = 1;
+		repeated string name_prefixes = 2;
+	}
+
+	Filters filters = 1;
+}
+
+// ListSecretResponse contains a list of all the secrets that match the name or
+// name prefix filters provided in `ListSecretRequest`.  The `SecretData.Spec.Data`
+// field in each `SecretData` object in each `Secret` should be nil instead of
+// actually containing the secret bytes.
+message ListSecretsResponse {
+	repeated Secret secrets = 1;
+}
+
+// CreateSecretRequest specifies a new secret (it will not update an existing
+// secret) to create.
+message CreateSecretRequest {
+	SecretDataSpec spec = 1;
+}
+
+// CreateSecretResponse contains the newly created `Secret`` corresponding to the
+// name in `CreateSecretRequest`.  Since it's a new secret, there should just be
+// the one version, and hence only one `SecretData` object in the `Secret`.
+// The `SecretData.Spec.Data` field should be nil instead of actually
+// containing the secret bytes.
+message CreateSecretResponse {
+	Secret secret = 1;
+}
+
+// UpdateSecretRequest specifies a new secret version to add to an existing secret
+// of the same name.
+message UpdateSecretRequest {
+	SecretDataSpec spec = 1;
+}
+
+// UpdateSecretResponse contains the Secret corresponding to the name of the spec
+// provided in `UpdateSecretRequest`, and it should be updated with an additional
+// `SecretData` object containing said spec.  The `SecretData.Spec.Data` field in
+// each `SecretData` object in the `Secret` object should be nil instead of
+// actually containing the secret bytes.
+message UpdateSecretResponse {
+	Secret secret = 1;
+}
+
+// RemoveSecretRequest contains the name and version (optional) of the secret
+// that should be removed.  If no version is provided, the intention is that
+// all versions should be removed.
+message RemoveSecretRequest {
+	// Name is name of the secret to remove.
```

Comment:
We only remove things by `id` and `version`, usually. This has very nice consistency guarantees. Imagine if someone removed `foo` and created a new `foo` in the time that another actor decided to remove `foo`. Using the `id` and `version` makes this operation safe.